# 1. Introduction

## 1.1 Problem Statement

- Rapido operates a large-scale ride-hailing platform, handling millions of daily bookings across multiple cities, vehicle types, and demand conditions. 
- Despite having rich, trip-level data—including bookings, customers, drivers, locations, and time signals—the company faces several operational challenges that impact business performance and customer satisfaction:

    + High Ride Cancellations: Frequent cancellations disrupt driver schedules, reduce revenue, and harm user experience.
    + Inaccurate Fare Estimation: Erroneous fare predictions lead to pricing inefficiencies and customer dissatisfaction.
    + Inefficient Driver Allocation: Poorly optimized allocation reduces operational efficiency and increases delays.
    + Unanticipated Operational Risks: Lack of predictive insights results in missed opportunities to proactively manage peak demand and high-risk rides.

- Objective:
- The goal of this project is to leverage Rapido’s large-scale booking data to build a unified, Machine Learning–driven decision system that:

    + Predicts ride outcomes (completed, cancelled, incomplete) before trip start.
    + Estimates fares dynamically and accurately based on trip and demand signals.
    + Identifies high-risk customers and unreliable drivers.
    + Enables data-driven operational interventions to improve customer experience, optimize driver allocation, and increase revenue.

- Expected Impact:

    + Reduce ride cancellations by at least 20%.
    + Improve ETA accuracy and operational efficiency.
    + Enable dynamic pricing based on demand prediction.
    + Provide actionable insights for driver reliability scoring and risk mitigation.

## 1.2 Business Usecases

- **Use Case 1 — Reduce Cancellations by 20%**

- Problem: High ride cancellations result in lost revenue, wasted driver availability, and poor customer experience.
- Solution:
    + Predict the probability of a ride being cancelled before it starts using historical booking, customer, and driver behavior data.
    + Identify high-risk customers and rides to take preventive measures (e.g., reminder notifications, incentivizing completion).
- Impact:
    + Minimize cancellation rates.
    + Optimize driver allocation and reduce idle time.

- **Use Case 2 — Improve ETA Accuracy**

- Problem: Inaccurate Estimated Time of Arrival (ETA) affects customer satisfaction and operational efficiency.
- Solution:

    + Predict driver delays based on historical performance, traffic patterns, and ride acceptance behavior.
    + Adjust ETAs dynamically in the app and notify customers proactively.
- Impact:
    + Enhance customer trust and satisfaction.
    + Reduce complaints and improve operational reliability.

- **Use Case 3 — Dynamic Pricing (Demand Prediction)**

- Problem: Static pricing models fail to account for peak demand, traffic, and weather conditions.
- Solution:

    + Build demand prediction models using time, location, traffic, and weather signals.
    + Implement dynamic fare estimation and surge pricing accordingly.
- Impact:
    + Maximize revenue during high-demand periods.
    + Ensure fair pricing and efficient resource allocation.

**Use Case 4 — Driver Reliability Scoring**

- Problem: Unreliable drivers lead to cancellations, delays, and poor service quality.
- Solution:

    - Score drivers based on historical acceptance rates, delay history, and customer feedback.
    - Allocate rides preferentially to high-reliability drivers.
- Impact:
    - Reduce incomplete rides and delays.
    - Improve overall ride quality and customer satisfaction.

## 1.3 Project Objectives

- The primary objective of this project is to develop an end-to-end data-driven system for analyzing ride-hailing operations and building predictive models to improve business outcomes. 
- The project is structured into the following key stages:

1. Data Profiling
    - Understand the structure, size, and quality of datasets (bookings, customers, drivers, etc.)
    - Identify:
        + Missing values
        + Data types
        + Outliers
        + Unique value distributions
        + Generate summary statistics to assess data readiness
2. Data Cleaning
    - Handle missing and inconsistent values
    - Remove duplicates and invalid records
    - Convert data types (e.g., timestamps to datetime)
    - Standardize categorical values
3. Exploratory Data Analysis (EDA)
    - Analyze patterns and trends in ride data:
    - Ride volume by time (hour, weekday, city)
    - Cancellation trends and peak hours
    - Distance vs fare relationship
    - Customer vs driver behavior
    - Identify key factors influencing cancellations, delays, and pricing
4. Feature Engineering
    - Create meaningful features to improve model performance:
    - Fare_per_KM, Fare_per_Min
    - Rush_Hour_Flag, Weekend_Flag
    - City_Pair (Pickup + Drop)
    - Driver_Reliability_Score
    - Customer_Loyalty_Score
    - Transform raw data into actionable signals
5. Data Preprocessing
    - Encode categorical variables (One-Hot / Label Encoding)
    - Normalize or scale numerical features
    - Handle class imbalance (if present)
    - Split dataset into training and testing sets (80/20)
6. Model Training
    - Build multiple Machine Learning models:
        + Classification (Ride Outcome, Cancellation Risk, Driver Delay)
        + Regression (Fare Prediction)
    - Use algorithms such as:
        + Logistic Regression
        + Random Forest
        + Gradient Boosting (XGBoost/LightGBM)
        + Perform hyperparameter tuning (GridSearch / Optuna)
7. Prediction & Evaluation
    - Evaluate model performance using:
        + Classification: Accuracy, F1-score, AUC, Confusion Matrix
        + Regression: RMSE, MAE, R²
    - Compare models and select the best-performing ones
    - Validate against industry benchmarks
8. Data Visualization (Streamlit Dashboard)
    - Build an interactive dashboard using Streamlit to:
    - Visualize ride trends and cancellations
    - Display model predictions (fare, cancellation risk)
    - Show driver and customer insights
    - Enable real-time, data-driven decision-making
    - Final Goal



# 2. Import Required Libraries

In [20]:
import sys
import os
project_root = os.path.abspath(os.path.join(os.getcwd(), ".."))  # one level up
# Add src folder to path
sys.path.append(os.path.join(project_root, "src"))
from data_loader import load_raw_data
import pandas as pd

In [21]:
# Set pandas display options to show all columns
pd.set_option('display.max_columns', None)

# 3. Load Data

In [22]:
bookings_df, customers_df, drivers_df, location_demand_df, time_features_df = load_raw_data()


# 4. Data Profiling

In [23]:
bookings_df.head()

,booking_id,booking_date,booking_time,day_of_week,is_weekend,hour_of_day,city,pickup_location,drop_location,vehicle_type,ride_distance_km,estimated_ride_time_min,actual_ride_time_min,traffic_level,weather_condition,base_fare,surge_multiplier,booking_value,booking_status,incomplete_ride_reason,customer_id,driver_id
0,B_000001,2025-12-11,00:07:00,Thursday,0,0,Mumbai,Loc_19,Loc_16,Bike,7.01,46.30,NaN,High,Heavy Rain,76.12,2.0,148.22,Cancelled,NaN,C_005097,D_004592
1,B_000002,2025-07-07,06:13:00,Monday,0,6,Mumbai,Loc_32,Loc_38,Cab,9.67,43.54,42.28,Medium,Heavy Rain,254.15,1.8,465.85,Completed,NaN,C_008459,D_000148
2,B_000003,2025-08-23,08:53:00,Saturday,1,8,Chennai,Loc_28,Loc_1,Auto,16.18,48.55,NaN,Low,Heavy Rain,234.20,1.9,457.03,Cancelled,NaN,C_003471,D_004976
3,B_000004,2025-04-12,10:25:00,Saturday,1,10,Delhi,Loc_16,Loc_30,Bike,1.02,4.61,4.76,Medium,Rain,28.20,1.8,51.03,Completed,NaN,C_002161,D_001173
4,B_000005,2025-08-23,00:08:00,Saturday,1,0,Hyderabad,Loc_22,Loc_31,Bike,12.35,55.56,64.53,Medium,Clear,118.77,1.2,144.73,Completed,NaN,C_005617,D_001175


In [24]:
customers_df.head()

,customer_id,customer_gender,customer_age,customer_city,customer_signup_days_ago,preferred_vehicle_type,total_bookings,completed_rides,cancelled_rides,incomplete_rides,cancellation_rate,avg_customer_rating,customer_cancel_flag
0,C_000001,Non-Binary,56,Bangalore,556,Cab,10,9,1,0,0.100000,3.8,0
1,C_000002,Male,46,Bangalore,82,Bike,8,7,1,0,0.125000,4.1,0
2,C_000003,Female,32,Delhi,969,Bike,7,4,1,2,0.142857,3.8,0
3,C_000004,Female,60,Hyderabad,549,Bike,11,7,3,1,0.272727,3.7,1
4,C_000005,Non-Binary,25,Hyderabad,359,Bike,9,5,3,1,0.333333,3.8,1


In [25]:
drivers_df.head()

,driver_id,driver_age,driver_city,vehicle_type,driver_experience_years,total_assigned_rides,accepted_rides,incomplete_rides,delay_count,acceptance_rate,delay_rate,avg_driver_rating,avg_pickup_delay_min,driver_delay_flag
0,D_000001,39,Bangalore,Auto,1,25,16,0,0,0.64,0.00,4.1,2.0,0
1,D_000002,40,Chennai,Cab,2,14,11,3,2,0.79,0.14,4.8,6.1,1
2,D_000003,26,Bangalore,Auto,12,19,14,3,2,0.74,0.11,4.1,2.6,1
3,D_000004,46,Chennai,Auto,3,18,13,2,2,0.72,0.11,4.2,2.6,1
4,D_000005,32,Chennai,Bike,7,18,16,1,1,0.89,0.06,4.8,4.9,0


In [26]:
location_demand_df.head()

,city,pickup_location,hour_of_day,vehicle_type,total_requests,completed_rides,cancelled_rides,avg_wait_time_min,avg_surge_multiplier,demand_level
0,Bangalore,Loc_1,0,Auto,2,2,0,87.940000,1.400000,Low
1,Bangalore,Loc_1,0,Bike,5,5,0,68.088000,1.460000,Low
2,Bangalore,Loc_1,0,Cab,6,5,0,50.913333,1.733333,Medium
3,Bangalore,Loc_1,1,Auto,3,2,0,72.883333,1.566667,Low
4,Bangalore,Loc_1,1,Bike,7,6,1,33.374286,1.242857,Medium


In [27]:
time_features_df.head()

,datetime,hour_of_day,day_of_week,is_weekend,is_holiday,peak_time_flag,season
0,2025-01-01 00:00:00,0,Wednesday,0,0,0,Winter
1,2025-01-01 01:00:00,1,Wednesday,0,0,0,Winter
2,2025-01-01 02:00:00,2,Wednesday,0,0,0,Winter
3,2025-01-01 03:00:00,3,Wednesday,0,0,0,Winter
4,2025-01-01 04:00:00,4,Wednesday,0,0,0,Winter


## 4.1 Basic Structure Overview

In [28]:
datasets = {
    "Bookings": bookings_df,
    "Customers": customers_df,
    "Drivers": drivers_df,
    "Location Demand": location_demand_df,
    "Time Features": time_features_df
}

In [29]:
for name, df in datasets.items():
    print(f"\n{name} Dataset")
    print("-" * 40)
    print("Shape:", df.shape)
    print("Columns:", df.columns.tolist())


Bookings Dataset
----------------------------------------
Shape: (100000, 22)
Columns: ['booking_id', 'booking_date', 'booking_time', 'day_of_week', 'is_weekend', 'hour_of_day', 'city', 'pickup_location', 'drop_location', 'vehicle_type', 'ride_distance_km', 'estimated_ride_time_min', 'actual_ride_time_min', 'traffic_level', 'weather_condition', 'base_fare', 'surge_multiplier', 'booking_value', 'booking_status', 'incomplete_ride_reason', 'customer_id', 'driver_id']

Customers Dataset
----------------------------------------
Shape: (10000, 13)
Columns: ['customer_id', 'customer_gender', 'customer_age', 'customer_city', 'customer_signup_days_ago', 'preferred_vehicle_type', 'total_bookings', 'completed_rides', 'cancelled_rides', 'incomplete_rides', 'cancellation_rate', 'avg_customer_rating', 'customer_cancel_flag']

Drivers Dataset
----------------------------------------
Shape: (5000, 14)
Columns: ['driver_id', 'driver_age', 'driver_city', 'vehicle_type', 'driver_experience_years', 'tota

- Booking dataset serves as the primary fact table for analysis and modeling(linking customers and drivers)
- Customer dataset is useful for customer risk profiling and cancellation prediction.
- Drivers dataset is used for demand forecasting and dynamic pricing.
- Time Features dataset capture temporal patterns and seasonality effects.
- One customer can make multiple bookings, but each booking belongs to one customer.
- One driver can handle multiple rides, but each booking has one assigned driver.

## 4.2 Data Types & Missing Values

In [30]:
for name, df in datasets.items():
    print(f"\n{name} - Info")
    print("-" * 40)
    print(df.info())


Bookings - Info
----------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 22 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   booking_id               100000 non-null  object 
 1   booking_date             100000 non-null  object 
 2   booking_time             100000 non-null  object 
 3   day_of_week              100000 non-null  object 
 4   is_weekend               100000 non-null  int64  
 5   hour_of_day              100000 non-null  int64  
 6   city                     100000 non-null  object 
 7   pickup_location          100000 non-null  object 
 8   drop_location            100000 non-null  object 
 9   vehicle_type             100000 non-null  object 
 10  ride_distance_km         100000 non-null  float64
 11  estimated_ride_time_min  100000 non-null  float64
 12  actual_ride_time_min     68346 non-null   float64
 13  tr

**Key Observations**
- The Bookings dataset contains:
    + 14 categorical (object) columns
    + 6 numerical (float) columns
    + 2 integer columns
- Columns such as:
    + booking_date
    + booking_time
    + are currently stored as object types and need conversion to datetime format.
- The Customers and Drivers datasets are:
    + Well-structured
    + Contain mostly numerical and categorical features
    + No immediate type issues
- The Time Features dataset:
    + Contains datetime as object type
    + Should be converted to proper datetime format for time-based analysis
- Most datasets are clean and structured
- Minor preprocessing required:
    + Datetime conversion
    + Categorical encoding (later stage)
    

## 4.3 Missing Value Analysis

In [31]:
for name, df in datasets.items():
    print(f"\n{name} - Missing Values")
    print("-" * 40)
    print(df.isnull().sum())


Bookings - Missing Values
----------------------------------------
booking_id                     0
booking_date                   0
booking_time                   0
day_of_week                    0
is_weekend                     0
hour_of_day                    0
city                           0
pickup_location                0
drop_location                  0
vehicle_type                   0
ride_distance_km               0
estimated_ride_time_min        0
actual_ride_time_min       31654
traffic_level                  0
weather_condition              0
base_fare                      0
surge_multiplier               0
booking_value                  0
booking_status                 0
incomplete_ride_reason     91630
customer_id                    0
driver_id                      0
dtype: int64

Customers - Missing Values
----------------------------------------
customer_id                 0
customer_gender             0
customer_age                0
customer_city               0
cust

**Key Findings**
- **Bookings Dataset** (Critical Insights)
    + actual_ride_time_min → 31,654 missing values (~31.6%)
    + incomplete_ride_reason → 91,630 missing values (~91.6%)
- These are NOT random missing values:
    + actual_ride_time_min
- Missing for:
    + Cancelled rides
    + Incomplete rides
- Logical because ride never finished
- Missing values are structural (business-driven), not data quality issues.
- **Customers Dataset**
    + No missing values
    + Fully clean and ready for modeling
- **Drivers Dataset**
    + No missing values
    + Reliable performance data
- **Location Demand Dataset**
    + No missing values
    + Clean aggregated dataset
- **Time Features Dataset**
    + No missing values
    + Ready for time-based feature engineering
- No imputation required for:
    + incomplete_ride_reason
- Conditional handling required for:
    + actual_ride_time_min

## 4.4 Duplicate Check

In [32]:
for name, df in datasets.items():
    print(f"\n{name} - Duplicates:", df.duplicated().sum())


Bookings - Duplicates: 0

Customers - Duplicates: 0

Drivers - Duplicates: 0

Location Demand - Duplicates: 0

Time Features - Duplicates: 0


In [33]:
bookings_df['booking_id'].nunique() == bookings_df.shape[0]

True

In [34]:
customers_df['customer_id'].nunique() == customers_df.shape[0]

True

In [35]:
drivers_df['driver_id'].nunique() == drivers_df.shape[0]

True

- Bookings:
    + Each booking_id is unique → no duplicate transactions
- Customers:
    + Each customer_id represents a unique customer
- Drivers:
    + Each driver_id represents a unique driver

- validated primary key uniqueness for bookings, customers, and drivers, ensuring one-to-many relationships and preventing duplication during joins.

- Even though keys are unique:
    + You can STILL get data inflation from Location Demand joins
    + Because that table depends on composite keys, not a single ID

## 4.5 Statistical Summary

### 4.5.1 Numerical Summary (All Datasets)

In [36]:
for name, df in datasets.items():
    print(f"\n{name} - Numerical Summary")
    print("-" * 50)
    print(df.describe())


Bookings - Numerical Summary
--------------------------------------------------
          is_weekend   hour_of_day  ride_distance_km  estimated_ride_time_min  \
count  100000.000000  100000.00000     100000.000000            100000.000000   
mean        0.284270      11.50605         13.028289                61.384020   
std         0.451068       6.91998          6.928495                39.299418   
min         0.000000       0.00000          1.000000                 3.000000   
25%         0.000000       5.00000          7.040000                29.887500   
50%         0.000000      12.00000         13.030000                55.430000   
75%         1.000000      17.00000         19.032500                86.060000   
max         1.000000      23.00000         25.000000               164.980000   

       actual_ride_time_min      base_fare  surge_multiplier  booking_value  
count          68346.000000  100000.000000     100000.000000  100000.000000  
mean              61.447278     2

**Bookings Dataset Insights**
- **Ride & Time Patterns**
    + Average ride distance: ~13 km
    + Average estimated ride time: ~61 mins
    + Wide variation:
        + Distance: 1 km → 25 km
        + Time: 3 min → 164 min

    + Indicates:
        + Mix of short and long-distance rides
        + Suitable for fare prediction modeling
**Pricing Insights**
- Average base fare: ~212
- Average booking value: ~336
- Surge multiplier range: 1.0 → 2.3

- Strong indication of:
    + Dynamic pricing behavior
    + Surge significantly impacts final fare

- actual_ride_time_min has fewer records → aligns with:
    + Cancelled rides
    + Incomplete rides

- Outlier Signal
    + Max booking value: 1265
    + Max ride time: 197 min

- Indicates:
    + Presence of extreme rides / peak surge scenarios

**Customers Dataset Insights**
- Customer Behavior
    + Avg bookings per customer: 10
    + Avg cancellation rate: ~23%
- High cancellation tendency → key business problem
- Ratings
    + Avg rating: ~4.25
    + Range: 3.5 → 5
- Customers are generally:
    + Satisfied but not perfect
- Key Insight
    + Cancellation rate varies up to 100%
- Strong signal for:
    + Customer risk prediction model

**Drivers Dataset Insights**
- Driver Performance
    + Avg acceptance rate: ~76%
    + Avg delay rate: ~4.7%
- Indicates:
    + Moderate reliability across drivers
- Delay Behavior
    + Avg pickup delay: ~3.2 minutes
    + Max delay rate: 42%
- Some drivers are:
    + Highly unreliable → critical for scoring
- Driver Ratings
    + Avg rating: ~4.49
    + Range: 4.0 → 5.0
- Drivers are generally:
    + High quality

**Location Demand Insights**
- Demand Patterns
    + Avg requests per slot: ~5.5
    + Max requests: 15

- Indicates:
    + Moderate but fluctuating demand
- Surge & Wait Time
    + Avg wait time: ~61 mins
    + Surge multiplier: 1.0 → 2.3

- Strong correlation likely between:
    + High demand → higher surge → longer wait

**Time Features Insights**
- Time Distribution
    + Uniform distribution across hours (0–23)
    + Peak time flag present (~29%)
- Dataset is:
    + Well-balanced across time
- Important Observation
    + is_holiday = 0 for all rows

- Means:
    + No holiday effect captured → potential limitation

- Critical Business Signals Identified
    + High customer cancellation rate (~23%)
    + Significant price variation due to surge
    + Presence of unreliable drivers
- Demand-driven variability in:
    + Wait time
    + Pricing

### 4.5.2 Categorical Summary (All Datasets)

In [37]:
for name, df in datasets.items():
    print(f"\n{name} - Categorical Summary")
    print("-" * 50)
    print(df.describe(include='object'))


Bookings - Categorical Summary
--------------------------------------------------
       booking_id booking_date booking_time day_of_week    city  \
count      100000       100000       100000      100000  100000   
unique     100000          365         1440           7       5   
top      B_099961   2025-04-06     23:58:00   Wednesday   Delhi   
freq            1          322           98       14558   20161   

       pickup_location drop_location vehicle_type traffic_level  \
count           100000        100000       100000        100000   
unique              50            50            3             3   
top             Loc_17         Loc_5          Cab          High   
freq              2126          2080        33415         33667   

       weather_condition booking_status incomplete_ride_reason customer_id  \
count             100000         100000                   8370      100000   
unique                 3              3                      4       10000   
top        

**Bookings Dataset – Key Insights**
| Column                          | Unique Values | Most Frequent           | Insights                                                                       |
| ------------------------------- | ------------- | ----------------------- | ------------------------------------------------------------------------------ |
| city                            | 5             | Delhi                   | Data spans 5 cities; Delhi has the most rides                                  |
| vehicle_type                    | 3             | Cab                     | Cab is most popular, followed by Bike and Auto                                 |
| traffic_level                   | 3             | High                    | High traffic occurs most often; relevant for ETA/fare models                   |
| weather_condition               | 3             | Heavy Rain              | Could affect cancellations, ride time                                          |
| booking_status                  | 3             | Completed               | ~68% of rides completed, others are Cancelled/Incomplete                       |
| pickup_location / drop_location | 50 each       | Loc_17 / Loc_5          | Popular pickup/drop points could affect surge & wait time                      |
| incomplete_ride_reason          | 4             | Driver Delay            | Most incomplete rides are due to driver delays                                 |
| customer_id                     | 10000         | C_009271 (max freq: 26) | Some customers book frequently — useful for loyalty/cancellation risk modeling |
| driver_id                       | 5000          | D_003592 (freq: 38)     | Frequent drivers can influence ETA, reliability scoring                        |


- Takeaway: Most categorical features have manageable cardinality, except IDs (which you’ll use for joins, not modeling).

**Customers Dataset – Key Insights**
| Column                 | Unique Values | Most Frequent | Insights                                                               |
| ---------------------- | ------------- | ------------- | ---------------------------------------------------------------------- |
| customer_gender        | 3             | Male          | Mostly male customers; can be a feature if gender affects cancellation |
| customer_city          | 5             | Mumbai        | Matches ride cities                                                    |
| preferred_vehicle_type | 3             | Bike          | Could influence vehicle allocation / cancellation rate                 |

- Takeaway: Useful for customer segmentation, cancellation risk, and loyalty features.

**Drivers Dataset – Key Insights**

| Column       | Unique Values | Most Frequent | Insights                                      |
| ------------ | ------------- | ------------- | --------------------------------------------- |
| driver_city  | 5             | Mumbai        | Matches ride cities                           |
| vehicle_type | 3             | Auto          | Useful for driver allocation & ETA            |
| driver_id    | 5000          | D_005000      | Unique, used for joins or scoring reliability |

- Takeaway: Driver features are categorical but low-cardinality — perfect for one-hot encoding or embeddings.

**Location Demand Dataset – Key Insights**

| Column          | Unique Values | Most Frequent | Insights                                                            |
| --------------- | ------------- | ------------- | ------------------------------------------------------------------- |
| city            | 5             | Chennai       | High demand location                                                |
| pickup_location | 50            | Loc_21        | Popular hotspots may influence surge, wait times                    |
| vehicle_type    | 3             | Cab           | Cab is dominant in location demand                                  |
| demand_level    | 2             | Low           | Binary demand classification; could be target for demand prediction |

- Takeaway: Perfect for temporal features (peak hours, weekends, seasonality).

In [38]:
categorical_columns = {
    'Bookings': ['day_of_week','city','pickup_location','drop_location',
                 'vehicle_type','traffic_level','weather_condition','booking_status','incomplete_ride_reason'],
    
    'Customers': ['customer_gender','customer_city','preferred_vehicle_type'],
    
    'Drivers': ['driver_city','vehicle_type'],
    
    'Location Demand': ['city','pickup_location','vehicle_type','demand_level'],
    
    'Time Features': ['day_of_week','season']
}

# Loop through each dataset and column
for ds_name, df in datasets.items():
    print(f"\n=== {ds_name} Dataset Categorical Value Counts ===\n")
    for col in categorical_columns[ds_name]:
        print(f"--- Column: {col} ---")
        print(df[col].value_counts(dropna=False).head(10))  # Top 10 values
        print("\n")


=== Bookings Dataset Categorical Value Counts ===

--- Column: day_of_week ---
day_of_week
Wednesday    14558
Friday       14411
Thursday     14336
Sunday       14245
Saturday     14182
Monday       14159
Tuesday      14109
Name: count, dtype: int64


--- Column: city ---
city
Delhi        20161
Chennai      20044
Bangalore    19947
Mumbai       19943
Hyderabad    19905
Name: count, dtype: int64


--- Column: pickup_location ---
pickup_location
Loc_17    2126
Loc_7     2095
Loc_37    2094
Loc_29    2067
Loc_3     2064
Loc_19    2064
Loc_20    2053
Loc_4     2045
Loc_48    2045
Loc_35    2041
Name: count, dtype: int64


--- Column: drop_location ---
drop_location
Loc_5     2080
Loc_27    2063
Loc_19    2053
Loc_14    2049
Loc_34    2045
Loc_33    2044
Loc_44    2043
Loc_39    2043
Loc_47    2036
Loc_1     2034
Name: count, dtype: int64


--- Column: vehicle_type ---
vehicle_type
Cab     33415
Bike    33394
Auto    33191
Name: count, dtype: int64


--- Column: traffic_level ---
traffic_

1. Bookings Dataset
- booking_status: Most rides are Completed (~68%), ~23% Cancelled, ~8% Incomplete.
- incomplete_ride_reason: Most incomplete rides are due to Driver Delay. Huge proportion of NaN as expected for completed/cancelled rides.
- vehicle_type and traffic_level are well-balanced across categories.
- day_of_week & city: Distribution is fairly even, small skew toward Delhi and Wednesday.

2. Customers Dataset
- Gender distribution is almost balanced across Male, Female, Non-Binary.
- Most customers prefer Bike slightly over Cab and Auto.
- City-wise customer base is fairly balanced, no extreme skew.

3. Drivers Dataset
- Driver distribution across cities is fairly even.
- Vehicle types are slightly skewed: Auto > Cab > Bike.
- Will help in driver delay prediction and driver reliability scoring.

4. Location Demand Dataset
- Slightly more Low demand than Medium, showing peak/off-peak trends.
- Vehicle types in demand almost equal.
- Each pickup location seems uniformly represented.

5. Time Features Dataset
- Season distribution is mostly uniform: Summer slightly higher.
- Days of the week: Wednesday slightly more entries (likely due to how datetime was generated).